### DISCUSSION TITLE
Student-managed RadGrad accounts

### DISCUSSION THREAD

**Discussion by philipmjohnson** [radgrad/radgrad2#802] [General]
# The problem

I am thinking about the long-term sustainability of RadGrad, and it occurs to me that a major barrier to adoption is the administrative overhead of (a) creating student accounts, and (b) uploading STAR data each semester.  If we could figure out a way for each student to manage their own account, it would be significantly easier to keep a RadGrad installation going.

# Managing STAR data

Let's look at problem 2 first, right now, a RadGrad admin has to login (from inside UH) each semester, get all of the emails from the RadGrad instance, and run a script to query STAR for the data.  Also, that admin needs to have special permission to obtain data about students. 

Now, let's imagine a checklist item that appears once a semester to step the student through a process which results in their STAR data getting uploaded to their RadGrad account.  I can imagine two approaches:

  1. Press a button on a RadGrad page. This pops up a screen requesting UH credentials, which are then used to login to the user's STAR account and scrape the appropriate pages.  (This seems complicated and brittle).
 
  2. Press a button on a RadGrad page. This pops up a screen requesting UH credentials, which are then used to issue a n HTTP request to STAR to return that user's course info.  (This requires help from the STAR team to provide that API access.)
 
  3. Press a button on a RadGrad page. This pops up a page of instructions on how to upload STAR data. This involves (a) logging into STAR on a separate browser window, navigating to a specific page, clicking the "download" button to get the data on the student's computer, then clicking an "upload" button on the student's RadGrad page to upload that data into their account. (This is highest overhead for students, but perhaps the STAR "download" button already exists? Also, it creates the possibility of students "spoofing" their course data by manually editing their downloaded course data before uploading, but I don't see any real benefit to the student for doing that.)

What do folks think about this idea? If you have student STAR access, can you investigate to see if there are currently ways to download your course history from STAR? 

# Account creation

If we have a way for students to manage their STAR data, then it becomes possible for us to let students register themselves in the system.

The big issue is that we want to restrict usage of a RadGrad instance not just to any enrolled UH student, but more specifically to those who are taking ICS courses.

One way to do this is to allow any student at UH to login to a RadGrad instance, but then set a timer and give them only 24 hours (say) to complete the registration process by uploading STAR data.  If they don't upload data within 24 hours, and if that data does not include any ICS courses, then the system puts a "lock" on their account and if they try to login again, they go to a page saying that they are not recognized as an ICS student (or whatever department associated with the instance) and that an admin (or advisor) must unlock their account for them to proceed.   

If an admin unlocks their account, then they have a new 24 hour period to get their ICS data uploaded. 

# Comments?

Please reply with your thoughts on this idea.

**Comment by cammoore** [comment]
## Managing STAR data

1. I don't like this solution at all. It is very difficult to create good scrapers for internships and we'll have to keep up to date with changes in STARs look and feel.
2. This seems like a security issue that the STAR folks may not want to open up. The connections to STAR will be on different IPs since we'll have to support different instances of RadGrad running on different servers.
3. This seems like the best solution if the students are able to download their STAR data. Just don't know if they have that functionality.

Having students manage their STAR data is a step in the right direction. I'm worried about the students who don't do it. Do we still want the admin to periodically update all the students? We're using this process to detect alumni. Students who have graduated won't update their STAR data and we won't detect that they have graduated.

## Account creation
We'd need a way to create/register an account on RadGrad. Not just log in with CAS. For the log-in to work we need the StudentProfile to exist. 

The students could click the register option. Then fill out a form to get their name. When they click submit we have them log in via CAS. If successful we'd create a StudentProfile and send them to another page indicating that they need to upload their STAR data within the 24-hour time limit.

The creation of the StudentProfile might be tricky since the user doing the registration process isn't logged in as Admin. I don't think we want the registration process to require an admin approval/creation of the StudentProfile.

**Comment by CalianaFortin** [comment]
### Managing Star Data
I like the idea of students being able to load their class data from STAR to RadGrad. 

1. This approach can become complicated. Like professor Moore said scraping can get difficult especially with pages that are volatile. (I don't think STAR changes their css much? IDK) I guess it would be a question of how often should we update the STAR scraper? Should it be constantly kept up to date, or should it only be updated when it needs to be?

| PROS  | CONS  |
|---|---|
| Scraping is something we all have experience with.  | Scraping can get difficult if the page is constantly changing.   |
| Simple way for the student download their data to the Radgrad system.   |   |

2. For me personally I like being about to only have to press one button to download my data. However, like professor Moore mentions there could be a security risk, with going about it this way. 

| PROS  | CONS  |
|---|---|
| Simple way for the student download their data to the Radgrad system.   | Security becomes and issue.  |

3. I think this option is better than the first option, but there are a few things I am worried about: 
- Is it too complicated for incoming freshmen? 
- Could it be complicated for students who are not tech savvy? 

| PROS  | CONS  |
|---|---|
| Gives students complete control of downloading/uploading their own data | Could potentially be a little complicated for incoming freshmen or students who are not tech savvy  |


From my knowledge students can download a PDF of their campus transcripts. This how the page looks:

<img width="1172" alt="Screen Shot 2021-10-29 at 3 16 34 PM" src="https://user-images.githubusercontent.com/70739998/139515152-a5f0fe24-68ac-4b7a-ac8b-d230fe753e10.png">

The Transcripts are formatted as such:

<img width="448" alt="Screen Shot 2021-10-29 at 3 37 16 PM" src="https://user-images.githubusercontent.com/70739998/139515884-4ed4eaf3-f9d7-4a25-82af-32219f4b1558.png">

I think it could be potentially beneficial for students to Manage STAR data with their advisor during mandatory advising. (IDK if this is a good idea? Just a thought.)

### Account creation
I personally think this idea is fine. It allows students to create an account at the time of their own choosing (Lets them decide if they want to use the system or not).

**Comment by jatinp101** [comment]
### **Managing STAR Data**

Overall I think it is an okay idea for students to manage their STAR data. I believe out of the 3 options you gave the third option to have students upload their STAR data up to RadGrad would be great due to 1 possibly being too difficult to do as Cam said " we'll have to keep up to date with changes in STARs look and feel." 2 wouldn't be a great idea either because as I work at STAR I have a feeling this probably would be something that I don't think they would be open to.  Catalina pointed out that this could be confusing for first-year students and students who aren't tech-savvy. I believe the best way we could mitigate this is to have a tutorial on the menu like a short video or gif showing how to do it. 

### **Account Creation**

I think this is a good idea. Personally, I would prefer to have to log in through CAS first and then be sent to the registration page to be able to insert my details in there, like how you would register for a site with a Google account (I hope that makes sense, if you would like me to elaborate more I can). We could probably have that check when logging in where if studentProfile doesn't exist we send them to the registration page. I think the 24/hour period seems pretty cumbersome and I think what would be better is making the upload of STAR data a required field when registering (if this can be done).

**Comment by saitoshi** [comment]
<h2>Managing STAR Data</h2>
I think it would  be interesting to have students manage their Star data but as Professor Moore mentioned it does bring in security concerns with the people at STAR that I do not think they would open up to it either for multiple reasons.  
<ol>
<li> While I do not think that the CSS Star constantly changes, I have noticed some minor changes time to time. Then the question of how often should we update teh scrapper Star comes into question since we also have other scrappers to work with such as Indeed, LinkedIn, etc.</li>
<li> Pressing a single button does make things easier compared to pressing multiple buttons. </li>
<li> While I doubt this is the case for the majority, as Caliana has mentioned it, the guide for those who are not technologically advanced or in Freshmen pressure should be super clear that way they are not pressured to not go away from RadGrad</li>
</ol>
<h2>Account Creation </h2>
I think this is a good idea as it gives students more option on whether they want to create an account or not.
